### PR TITLE
[DAT-2574] Callable Statements Documentation

### PR DESCRIPTION
**Pull request by snajahi**

### CODE REVIEW

File: content/appstore/connectors/database-connector.md [hunk 1]
```diff
@@ -35,25 +35,82 @@ These are the prerequisites for using this connector:
 
 ## 3 Getting Started
 
-### 3.1 Usage
+Once you have imported the Database Connector into your app, you will have the **Database Connector** category available in the **Toolbox**. The connector supports five actions: **Execute query**, **Execute statement**, **Execute parameterized query**, **Execute parameterized statement** and **Execute callable statement**.
```
Comment (dahfjkg) [inline content/appstore/connectors/database-connector.md:38]: minor: I would list the actions alphabetically.

File: content/appstore/connectors/database-connector.md [hunk 2]
```diff
@@ -35,25 +35,82 @@ These are the prerequisites for using this connector:
...
+#### 3.1.2 Results
 
 These are the results of the actions:
 
 * **Execute query** and **Execute parameterized query** – a list of objects of the row type, which is also the output of the `SELECT` SQL query
 * **Execute statement** and **Execute parameterized statement** – either an integer or a long value, which usually represents the amount of affected rows
 
+### 3.2 Executing Callable Statements
+
+The **Execute callable statement** microflow action is used to execute stored procedures. This action has no output and expects an input object of type `DatabaseConnector.Statement` which should defines the contract to call the stored procedure and receive the results:
```
Comment (dahfjkg) [inline content/appstore/connectors/database-connector.md:64]: - `is used to execute stored procedures` — this is not entirely correct. You can also call anonymous functions.
- `This action has no output` is a bit misleading since the Statement object can contain both input and output parameters. I would say something along the lines of "This action takes an object of type `DatabaseConnector.Statement` as an input. That object contains associations to input and output parameters of the call"

Comment (dahfjkg) [inline content/appstore/connectors/database-connector.md:64]: I would briefly mention the rest of the parameters here — credentials and JDBC URL

File: content/appstore/connectors/database-connector.md [hunk 3]
```diff
@@ -35,25 +35,82 @@ These are the prerequisites for using this connector:
...
+
+![](attachments/database/callable-statement-action.png)
+
+The `DatabaseConnector.Statement` type is a non-persistable entity defined as follows:
+
+![](attachments/database/statement-parameter-diagram.png)
+
+The **contents** attribute of the `DatabaseConnector.Statement` type should contain the statement body (i.e. the SQL content). An association with a `DatabaseConnector.Parameter` type is also available to define, if applicable, the input and output parameters that the stored procedure expects.
```
Comment (dahfjkg) [inline content/appstore/connectors/database-connector.md:72]: - `The **contents** attribute` — I would use exactly the same name as in the module, `Content`

Comment (dahfjkg) [inline content/appstore/connectors/database-connector.md:72]: It is not entirely clear from this description what is meant by statement body. I think we should include at least one example of usage somewhere in this guide.

File: content/appstore/connectors/database-connector.md [hunk 4]
```diff
@@ -35,25 +35,82 @@ These are the prerequisites for using this connector:
...
+
+![](attachments/database/callable-statement-action.png)
+
+The `DatabaseConnector.Statement` type is a non-persistable entity defined as follows:
+
+![](attachments/database/statement-parameter-diagram.png)
+
+The **contents** attribute of the `DatabaseConnector.Statement` type should contain the statement body (i.e. the SQL content). An association with a `DatabaseConnector.Parameter` type is also available to define, if applicable, the input and output parameters that the stored procedure expects. 
+
+The `DatabaseConnector.Parameter` type uses the **Name** and **Position** attributes to refer to either the parameter name or position in the stored procedure. It also has a **ParameterMode** attribute to indicate which parameter mode to use:
```
Comment (dahfjkg) [inline content/appstore/connectors/database-connector.md:74]: I would mention that `Name` and `Position` should not be used in the same statement: in one statement all parameters should be identified uniformly by name or uniformly by position.

File: content/appstore/connectors/database-connector.md [hunk 5]
```diff
@@ -35,25 +35,82 @@ These are the prerequisites for using this connector:
...
+##### a. Primitive Types
+The `DatabaseConnector.ParameterDecimal`, `DatabaseConnector.ParameterLong`, `DatabaseConnector.ParameterDateTime` and `DatabaseConnector.ParameterString` can be used to refer to the SQL primitive types for decimals, natural numbers, dates and character types respectively.
+
+![](attachments/database/primitive-types-parameters.png)
+
+The **Value** attribute defined in these specializations will be handled differently depending on the parameter mode used. In case of input parameters, the attribute must hold the value to pass to the stored procedure. In case of output parameters, it will be set to the output from the stored procedure.
+
+
+##### b. ParameterObject Type
+For complex SQL object types, the `DatabaseConnector.ParameterObject` can be used. The **SQLTypeName** attribute should be set to the underlying SQL type.
```
Comment (dahfjkg) [inline content/appstore/connectors/database-connector.md:94]: 1. I would mention that object type needs to be defined in the database before the statement is executed.
2. I wonder if we should mention that this is applicable only to some databases. As I understand, objects are available in Oracle and PostgreSQL, but in some databases there is no such concept.

File: content/appstore/connectors/database-connector.md [hunk 6]
```diff
@@ -35,25 +35,82 @@ These are the prerequisites for using this connector:
...
+
+![](attachments/database/parameter-object-parameter.png)
+
+The many-to-many association with `DatabaseConnector.Parameter` allows for flexibility to define any object hierarchy needed to accurately represent the underlying SQL typ, i.e. Objects with primitive attributes, objects with other nested objects etc.
```
Comment (dahfjkg) [inline content/appstore/connectors/database-connector.md:98]: I would not focus on the association type here because it is an implementation detail. In fact, it would be better to have a 1-many association from `ParameterObject` to `Parameter` owned by `ParameterObject`, but Mendix does not allow that.

I would rephrase this and say that attributes of the object can be defined as specializations of `DatabaseConnector.Parameter` using `ParameterObject_Parameter` association. The module supports all specializations as long as they are supported by the underlying database.

Comment (dahfjkg) [reply content/appstore/connectors/database-connector.md:98]: I would also mention that attributes within an object are identified by position.

File: content/appstore/connectors/database-connector.md [hunk 7]
```diff
@@ -35,25 +35,82 @@ These are the prerequisites for using this connector:
...
+
+##### c. ParameterList Type
+List parameters are also supported and usable via the `DatabaseConnector.ParameterList` type. The **SQLTypeName** attribute should refer to the SQL list name:
```
Comment (dahfjkg) [inline content/appstore/connectors/database-connector.md:101]: `SQLTypeName` should be a list type defined in the database beforehand, not name.

File: content/appstore/connectors/database-connector.md [hunk 8]
```diff
@@ -35,25 +35,82 @@ These are the prerequisites for using this connector:
...
+
+![](attachments/database/parameter-list-parameter.png)
+
+This type also defines a many-to-many association with `DatabaseConnector.Parameter` for flexibility purposes. Lists of primitive types, list of objects and list of lists should be possible to represent using this structure.
```
Comment (dahfjkg) [inline content/appstore/connectors/database-connector.md:105]: see my comments for Objects

File: content/appstore/connectors/database-connector.md [hunk 9]
```diff
@@ -35,25 +35,82 @@ These are the prerequisites for using this connector:
...
+ To use any of **Execute query**, **Execute statement**, **Execute parameterized query** or **Execute parameterized statement** in your Mendix application, drag them into your microflow. Next, provide all the arguments for the selected action and choose the output result name.
+ 
+ The **Execute query** and **Execute parameterized query** actions should be used for querying objects with a `SELECT` SQL command. The **Execute statement** and **Execute parameterized statement** actions should be used for all other commands (for instance, `INSERT`, `UPDATE`, or `DELETE`).
+ 
+ For both queries and statements, the difference between the parameterized and regular versions are that the parameterized version takes a string template parameter, while the regular version takes a fully formed SQL command string with no placeholders.
+ 
+ {{% alert type="info" %}}
+ The parameterized actions are only available with Database Connector versions 3.0.0 and above. For these, it is necessary to use Mendix [8.6.0](/releasenotes/studio-pro/8.6#860).
```
Comment (dahfjkg) [inline content/appstore/connectors/database-connector.md:52]: shall we update this block to mention callable statements? or we can remove it altogether since we support modules in 8.12 now

File: content/appstore/connectors/database-connector.md [hunk 10]
```diff
@@ -35,25 +35,82 @@ These are the prerequisites for using this connector:
...
 
-The **Execute query** and **Execute parameterized query** actions should be used for querying objects with a `SELECT` SQL command. The **Execute statement** and **Execute parameterized statement** actions should be used for all other commands (for instance, `INSERT`, `UPDATE`, or `DELETE`).
+### 3.1 Executing Queries And Statements
+#### 3.1.1 Usage
 
-For both queries and statements, the difference between the parameterized and regular versions are that the parameterized version takes a string template parameter, while the regular version takes a fully formed SQL command string with no placeholders.
+ To use any of **Execute query**, **Execute statement**, **Execute parameterized query** or **Execute parameterized statement** in your Mendix application, drag them into your microflow. Next, provide all the arguments for the selected action and choose the output result name.
+ 
+ The **Execute query** and **Execute parameterized query** actions should be used for querying objects with a `SELECT` SQL command. The **Execute statement** and **Execute parameterized statement** actions should be used for all other commands (for instance, `INSERT`, `UPDATE`, or `DELETE`).
```
Comment (dahfjkg) [inline content/appstore/connectors/database-connector.md:47]: - `all other commands` — I would be more specific here. Would it be correct to say `All DML commands` or are there other commands that can be executed this way?

File: content/appstore/connectors/database-connector.md [hunk 11]
```diff
@@ -35,25 +35,82 @@ These are the prerequisites for using this connector:
...
+##### d. ParameterRefCursor Type
+To deal with **REF CURSOR** outputs, use the `DatabaseConnector.ParameterRefCursor` type:
+
+![](attachments/database/parameter-ref-cursor-parameter.png)
+
+The many-to-many-association with `DatabaseConnector.Parameter` is also used here for the same reasons mentioned in the previous section about lists parameters.
+
 ## 4 Best Practices
 
 * Avoid having a user input as part of your dynamic SQL queries and statements (in the future, using parameters with queries or statements will be supported)
```
Comment (dahfjkg) [inline content/appstore/connectors/database-connector.md:116]: This seems outdated

**Review by dahfjkg — CHANGES_REQUESTED** [review]

**Comment by dahfjkg** [comment]
`content/appstore/connectors/attachments/database/Thumbs.db` does not need to be in this PR

**Comment by dahfjkg** [comment]
Passed the internal review. We will create another PR to add some examples later. @Adam-Dupaski

**Comment by Adam-Dupaski** [comment]
@dahfjkg if it's no longer in draft, please click Ready for Review to take it out of the draft state. Can this be merged right away after my review?

**Comment by Adam-Dupaski** [comment]
Thanks @snajahi - I will review and merge when done.

**Review by Adam-Dupaski — APPROVED** [review]
